Add tokio-buf and a `BufStream` trait

carllerche opened the pull request:
This is a **work in progress**.

The description will be updated over time.

# TODO

* [x] `SizeHint` guarantees
* [x] Should `consume_hint` take a `u64` or `usize`?
* [x] `BufStream::limit()` tests

carllerche commented:
@seanmonstar I have updated `size_hint` documentation to say that the hint values must be accurate. This should work for Hyper to use the hint for the `content-length` value?

carllerche commented:
The problem, however, is now `size_hint` is "useless" for gzip stream transformations. When compressing a stream, there is no guarantee as to the resulting size. The stream can even get larger, which means that an upper bound cannot be set.

carllerche commented:
Perhaps, `size_hint` should have an `approximate` field that is between upper / lower that is used to reserve memory when collecting the stream?

seanmonstar commented on tokio-buf/src/buf_stream/mod.rs:
```diff
@@ -0,0 +1,112 @@
+    ///
+    /// Once a stream is finished, i.e. `Ready(None)` has been returned, further
+    /// calls to `poll` may result in a panic or other "bad behavior".
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error>;
```
Should this be `poll_next` like the upcoming change to `Stream`? Or perhaps a different name, since a type could conceivably implement both `Stream` and `BufStream`, and if both traits are in scope, so there is no conflict?

seanmonstar commented on tokio-buf/src/buf_stream/mod.rs:
```diff
@@ -0,0 +1,112 @@
+    /// calls to `poll` may result in a panic or other "bad behavior".
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error>;
+
+    /// Returns the bounds on the remaining length of the iterator.
```
Calls this an "iterator"

seanmonstar commented on tokio-buf/src/buf_stream/size_hint.rs:
```diff
@@ -0,0 +1,79 @@
+pub struct SizeHint {
+    available: usize,
+    lower: usize,
+    upper: Option<usize>,
```
For use in hyper to set the content-length, this being a `usize` *does* mean that 32bit platforms can't set the content-length automatically if sending over 4GBs. Also, things like implementing `BufStream` for `tokio::fs::File` would cause problems there as well, since [`Metadata::len()`](https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.len) returns a `u64`.

carllerche replied:
Changing `size_hint` to `u64` has the interesting property of making `collect()` fallible.

carllerche commented:
@seanmonstar I switched `SizeHint` to have `u64` fields. Doing this exposed a flaw in `collect()` where large streams could result in overflow situations.

I resolved this by adding an error associated type to `FromBufStream`. See [here](https://github.com/tokio-rs/tokio/pull/611/commits/146877612a4e807f72d3cbc98fc2bcca17d927fd#diff-026fa19f4a16876a112c89b16c89bccaR21).

This ended up complicating the `FromBufStream` implementation for `Vec<u8>` as I added handling for overflow. See [here](https://github.com/tokio-rs/tokio/pull/611/commits/146877612a4e807f72d3cbc98fc2bcca17d927fd#diff-026fa19f4a16876a112c89b16c89bccaR67).

About half of those changes are handling potential overflow and the other half is factoring `SizeHint::upper()`.

carllerche commented:
Then the next question... should `consume_hint` be `u64` or `usize`?

carllerche commented:
I added a `limit` combinator [here](https://github.com/tokio-rs/tokio/pull/611/files#diff-e7207f5219ca9f7218d7b5398358e744R124).

Now, to collect a body stream, you can do:

```rust
let body = body_stream
    .limit(mb_100)
    .collect();
```

carllerche commented:
I think I'm going to get rid of `SizeHint::available` for now. It can be added later and it is not obviously needed.

seanmonstar commented on tokio-buf/src/buf_stream/mod.rs:
```diff
@@ -0,0 +1,168 @@
+    /// if it could produce data at that point. If it chooses to return
+    /// `NotReady`, when `consume_hint` is called with a non-zero argument, the
+    /// task must be notified in order to respect the `poll_buf` contract.
+    fn consume_hint(&mut self, amount: usize) {
```
A possible alternative to this function is passing some hint as an argument to `poll_buf`. Like, `stream.poll_buf(1024 * 16)` or something...

seanmonstar replied:
Is there an example of this providing value somewhere? I get the intention, but I'd personally prefer this left off until there's an implementation showing how it can be helpful.

seanmonstar commented:
Notably, removing hyper's optimization for [`Payload::__hyper_full_data`](https://github.com/hyperium/hyper/blob/dcf034715183e719ff8df0f1edb3a1d70b71f761/src/body/body.rs#L327) results in (on my  machine) ~9% reduction in [this pipeline benchmark](https://github.com/hyperium/hyper/blob/master/benches/pipeline.rs). This is a private-ish method that allows skipping checking the size-hint and then polling and then checking is_end_stream, opting for just a single method call, when the `Body` already has the full data available.

carllerche commented:
@seanmonstar what is the additional cost from?

carllerche commented:
@seanmonstar it seems highly suspect that using `BufStream` would be slower given that calling `size_hint` on `Once` *should* be extremely fast and you can use the size_hint to only poll once and skip is_end_stream.

seanmonstar commented:
So, some of the slow down I measured was that internally a optimized fast path was not being taken. Having refactored to basically this:

```rust
match body.poll_data()? {
    Async::Ready(Some(data)) => {
        if body.is_end_stream() {
            self.write_full_msg(head, data);
        }
    }
}
```

This has improved it somewhat, but there is still ~5% slowdown compared to when using `body.__hyper__full_body`. I suspect it's because when using `__hyper_full_body`, there's no branch to deal with an `Err` or `NotReady`, and no second check for `is_end_stream`.

carllerche commented:
@seanmonstar that is pretty surprising and would be worth digging into.
